code/go/0chain.net/chaincore/currency/currency.go

```diff
@@ -0,0 +1,181 @@
+	// ErrFloat64UnderflowsUint64 is returned if when converting an float6464 to a uint64 underflow uint64
+	ErrFloat64UnderflowsUint64 = errors.New("float64 underflows uint64")
+)
+
+var maxDecimal decimal.Decimal
+
+func init() {
+	maxDecimal = decimal.NewFromInt(math.MaxInt64)
+}
+
+//go:generate msgp -io=false -tests=false -v
+//Coin - any quantity that is represented as an integer in the lowest denomination
+type Coin uint64
+
+func ParseZCN(c float64) (Coin, error) {
+	d := decimal.NewFromFloat(c)
+	if d.Sign() == -1 {
+		return 0, ErrNegativeValue
+	}
+
+	// ZCN have a maximum of 10 decimal places
+	if d.Exponent() < -ZCNExponent {
+		return 0, ErrTooManyDecimals
+	}
+
+	// Multiply the coin balance by 1e10 to obtain coin amount
+	e := d.Shift(ZCNExponent)
+
+	// Check that there are no decimal places remaining. This error should not
+	// occur, because of the earlier check of ZCNExponent()
+	if e.Exponent() < 0 {
+		return 0, ErrTooManyDecimals
+	}
+
+	// Values greater than math.MaxInt64 will overflow after conversion to int64
+	if e.GreaterThan(maxDecimal) {
+		return 0, ErrTooLarge
+	}
+
+	return Coin(e.IntPart()), nil
+}
+
+func (c Coin) ToZCN() (float64, error) {
+	if c > math.MaxInt64 {
+		return 0.0, ErrTooLarge
+	}
+
+	f, _ := decimal.New(int64(c), -ZCNExponent).Float64()
+	return f, nil
+}
+
+// Int64 converts c uint64 Coin to an int64, returning an error if the uint64 value overflows int64
+func (c Coin) Int64() (int64, error) {
+	b := int64(c)
+	if b < 0 {
+		return 0, ErrUint64OverflowsInt64
+	}
+	return b, nil
+}
+
+// Float64 converts c uint64 Coin to a float64, returning an error if the uint64 value overflows float64
+func (c Coin) Float64() (float64, error) {
+	b := float64(c)
+	if b < 0 {
+		return 0, ErrUint64OverflowsFloat64
+	}
+	return b, nil
+}
+
+// MultUint64 multiplies Coin c by b, returning an error if the values overflow
+func MultUint64(c Coin, b uint64) (Coin, error) {
+	a := uint64(c) * b
+	if a != 0 && a/uint64(c) != b {
+		return 0, ErrUint64MultOverflow
+	}
+	return Coin(a), nil
+}
+
+// AddCoin adds c and b, returning an error if the values overflow
+func AddCoin(c, b Coin) (Coin, error) {
+	sum := c + b
+	if sum < c || sum < b {
+		return 0, ErrUint64AddOverflow
+	}
+	return sum, nil
+}
+
+// AddInt64 adds c and a, returning an error if the values overflow
+func AddInt64(c Coin, a int64) (Coin, error) {
+	b, err := Int64ToCoin(a)
+	if err != nil {
+		return 0, err
+	}
+	return AddCoin(c, b)
+}
+
+// MinusCoin subtracts b from c, returning an error if the values overflow
+func MinusCoin(c, b Coin) (Coin, error) {
+	if b > c {
+		return 0, ErrUint64MinusOverflow
+	}
+	sub := c - b
+	return sub, nil
+}
+
+// MinusInt64 subtracts a from c, returning an error if the values overflow
+func MinusInt64(c Coin, a int64) (Coin, error) {
+	b, err := Int64ToCoin(a)
+	if err != nil {
+		return 0, err
+	}
+	return MinusCoin(c, b)
+}
+
+func (c Coin) DivideCoin(a int64) (oCur, bal Coin, err error) {
```

Comment:
Sorry, I mean do this in a function instead of a method, i.e., `func DivideCoin(a Coin, b int64) (...)`